### ISSUE TITLE
CHPL_COMM_DEBUG and GASNET_MISC_CFLAGS

### ISSUE THREAD

**Issue by PHHargrove** [chapel-lang/chapel#7010]
I am trying to deploy some CI scripts to test Chapel's "master" with GASNet's "develop" on a daily basis.  While I am having reasonably good luck with default builds, I am not having much luck with getting CHPL_COMM_DEBUG to work as I am expecting.

This is the first (most pressing for me) issue I've encountered.

When building comm-gasnet.c, your Makefile logic appears to be *intended* to use `$(GASNET_MISC_CFLAGS)`, which is necessary to suppress various warnings in GASNet's headers that are otherwise fatal due to your use of `-Werror`.  However, my testing shows that is *not* happening.  I've traced the problem to the following logic in third-party/gasnet/Makefile.setup:
```
ifeq ($(CC),$(GASNET_CC))
CHPL_GASNET_CFLAGS_ALL = $(GASNET_MISC_CPPFLAGS) $(GASNET_DEFINES) $(GASNET_INCLUDES) $(GASNET_MISC_CFLAGS) $(MANUAL_CFLAGS) $(CHPL_GASNET_MORE_CFLAGS)
OPT_CFLAGS += $(GASNET_OPT_CFLAGS)
CHPL_GASNET_LDFLAGS = $(GASNET_LDFLAGS)
else
# If GASNet is using a different compiler from the rest of the Chapel build,
# don't use GASNet's
# GASNET_MISC_CPPFLAGS
# GASNET_MISC_CFLAGS
# GASNET_OPT_CFLAGS
# MANUAL_CFLAGS
CHPL_GASNET_CFLAGS_ALL = $(GASNET_DEFINES) $(GASNET_INCLUDES) $(CHPL_GASNET_MORE_CFLAGS)
endif
```

The issue lies within the very first line, which is testing for equality between your C compiler and GASNet's.  The issue is that `$(GASNET_CC)` is always a fully-qualified path such as `/usr/bin/gcc`, while `$(CC)` is almost always unqualified, such as `gcc`.

I tried working around this by building chapel with `gmake CC=/usr/bin/gcc` to force equality in that test.  However, that doesn't work because setting `CC` in this manner (for a gcc that defaults to C89) leads to a failure building third-party/qthread.  Here is just one of many error messages:
```
/tmp/BLD/dbg/chapel/third-party/qthread/qthread-src/src/feb.c: In function 'qt_feb_subsystem_shutdown':
/tmp/BLD/dbg/chapel/third-party/qthread/qthread-src/src/feb.c:112:5: error: 'for' loop initial declarations are only allowed in C99 mode
     for (unsigned i = 0; i < QTHREAD_LOCKING_STRIPES; i++) {
     ^
/tmp/BLD/dbg/chapel/third-party/qthread/qthread-src/src/feb.c:112:5: note: use option -std=c99 or -std=gnu99 to compile your code
```

I could try `CC="gcc -std=gnu99"`, but that is getting too far from the desire to have a system that "just works".

What I am currently doing is replacing
```
ifeq ($(CC),$(GASNET_CC))
```
with
```
ifeq ($(CC),$(notdir $(GASNET_CC)))
```

While this is not perfectly robust, it seemed easier to me that trying to get a full path for `$(CC)`.  
This solution at least handles the case of gcc vs clang, icc, etc..

**Comment by bradcray** [comment]
Thanks for pointing this out Paul.  I'm not familiar with the Makefile code in question (looks like @mppf developed it and that @ronawho's spent some time in there), but I wanted to note for those following up on this issue that we've also use our own Makefile variable `CHPL_GASNET_MORE_CFLAGS` to squash GASNet-specific warnings that would cause -Werror problems.  If the above were fixed, this variable could become redundant (?).

[Paul, you may be happy to hear that I removed -Wno-unused from this variable last week as a result of upgrading to GASNet 1.28.2).

**Comment by ronawho** [comment]
I'm not familiar with this code already (I just did the LFLAGS => LDFLAGS fix), so I'll let @mppf comment first

**Comment by PHHargrove** [comment]
@bradcray 
I had tried using `CHPL_GASNET_MORE_CFLAGS`, but gave up quickly when I realized that not only could I eventually need a different incantation for every compiler family, but that I additionally had to dive into `make/compiler/Makefile.*` to merge your values with my own (no way for me to add to the variable w/o replacing your settings).  

re: removal of `-Wno-unused`:  
That removal is correct for all versions of clang, and for gcc newer than some version I don't recall.    
For older gcc (including /usr/bin/gcc on RHEL/CentOS/ScientificLinux 7.3) the flag is still required and `$(GASNET_MISC_CFLAG)` is our recommended way to get it.  So, ironically I may not have noticed this problem as quickly if you had not removed `-Wno-unused`.

**Comment by bradcray** [comment]
Hi Paul -- 

Thinking about this a bit more after your most recent response, I'm surprised that you're having `-Werror`-related issues, as we typically only throw this if `CHPL_DEVELOPER` is set or `make WARNINGS=1` is used (i.e., "I'm someone who should care about code cleanliness").  Are you using either of these? (I'd guess not).  Yet you're seeing `-Werror` show up on compilations of `comm-gasnet.c`?

In any case, I agree that there's a fix needed here and didn't mean to suggest that `CHPL_GASNET_MORE_CFLAGS` would be a reasonable solution for you to take.

**Comment by PHHargrove** [comment]
@bradcray 

> [...] we typically only throw this if CHPL_DEVELOPER is set [...]

Yup.  You got 1 step ahead of me.  
As I just reported in #7011, I need `CHPL_DEVELOPER=1` to get past a second issue I've encountered with `CHPL_COMM_DEBUG=1` builds.

**Comment by mppf** [comment]
>  What I am currently doing is replacing
>  ```
>  ifeq ($(CC),$(GASNET_CC))
>  ```
>  with
>  ```
>  ifeq ($(CC),$(notdir $(GASNET_CC)))
>  ```
>
>  While this is not perfectly robust, it seemed easier to me that trying to get a full path for `$(CC)`.  
>  This solution at least handles the case of gcc vs clang, icc, etc..

That sounds good to me and matches the intent of the original reason this conditional was added (in PR #1866). Do you plan to make a PR with this change?

**Comment by cassella** [comment]
FWIW, I believe I'm hitting the "various warnings in GASNet's headers"
mentioned in the opening comment.  I'm using ``CHPL_DEVELOPER``, but
not ``CHPL_COMM_DEBUG``, ``CHPL_DEBUG``, ``DEBUG``, or ``OPTIMIZE``.

(The -Werrors are static functions such as ``_gasneti_threadkey_init``
being defined but not used while building ``comm-gasnet.c``.)

It builds with ``CHPL_DEVELOPER`` unset, or with the Makefile change
to add ``notdir``, above.

```
fortytwo@magrathea:~/src/chapel (master)$ printchplenv --anonymize
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: native *
CHPL_LOCALE_MODEL: flat
CHPL_COMM: gasnet *
  CHPL_COMM_SUBSTRATE: udp
  CHPL_GASNET_SEGMENT: everything
CHPL_TASKS: qthreads
CHPL_LAUNCHER: amudprun
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_MAKE: make
CHPL_ATOMICS: intrinsics
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: gmp
CHPL_HWLOC: hwloc
CHPL_REGEXP: re2
CHPL_WIDE_POINTERS: struct
CHPL_AUX_FILESYS: none

fortytwo@magrathea:~/src/chapel (master)$ echo $CHPL_DEVELOPER
1
fortytwo@magrathea:~/src/chapel (master)$ echo $CHPL_COMM_DEBUG $DEBUG $OPTIMIZE

fortytwo@magrathea:~/src/chapel (master)$ gcc --version
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.4) 5.4.0 20160609
```

**Comment by PHHargrove** [comment]
@mppf 
> That sounds good to me and matches the intent of the original reason this conditional was added (in PR #1866). Do you plan to make a PR with this change?

Alas I lack the necessary contributors agreement.

**Comment by PHHargrove** [comment]
@cassella 
Yes, it definitely sounds like you are encountering the same issue.

**Comment by bradcray** [comment]
I think that this issue can be closed with PR #7031 now merged, is that right @mppf?

**Comment by mppf** [comment]
@PHHargrove - I'm closing this issue but please ask to re-open it if you don't feel the problem has been addressed.
